adapter-base/src/main/java/org/eclipse/hono/adapter/auth/device/TenantServiceBasedX509Authentication.java

```diff
@@ -135,72 +147,113 @@ public TenantServiceBasedX509Authentication(
+                                        });
+                                        log.trace(b.toString());
+                                    }
+                                    final List<X509Certificate> chainToValidate = List.of(deviceCert);
+                                    return certPathValidator.validate(chainToValidate, trustAnchors)
+                                            .recover(t -> Future.failedFuture(new ClientErrorException(
+                                                    tenant.getTenantId(),
+                                                    HttpURLConnection.HTTP_UNAUTHORIZED,
+                                                    t.getMessage(),
+                                                    t)));
+                                }
+                            })
+                            .compose(ok -> getCredentials(x509chain, tenantTracker.result()));
+                })
+                .onSuccess(authInfo -> span.log("certificate validated successfully"))
+                .onFailure(t -> {
+                    log.debug("validation of client certificate failed", t);
+                    TracingHelper.logError(span, t);
+                })
+                .onComplete(ar -> span.finish());
     }
 
-    private Future<TenantObject> getTenant(final X509Certificate clientCert, final Span span) {
+    private Future<TenantObject> getTenant(
+            final X509Certificate clientCert,
+            final List<String> requestedHostNames,
+            final Span span) {
+
+        return tenantClient.get(clientCert.getIssuerX500Principal(), span.context())
+                .recover(t -> {
+                    final int statusCode = ServiceInvocationException.extractStatusCode(t);
+                    if (log.isDebugEnabled()) {
+                        log.debug("failed to look up tenant using trust anchor [subject DN: {}, status code: {}]",
+                                clientCert.getIssuerX500Principal().getName(X500Principal.RFC2253),
+                                statusCode);
+                    }
+                    span.log("failed to look up tenant using trust anchor");
+
+                    if (statusCode == HttpURLConnection.HTTP_NOT_FOUND && requestedHostNames.size() > 0) {
+
+                        final String hostName = requestedHostNames.get(0);
+                        final int idx = hostName.indexOf('.');
 
-        return tenantClient.get(clientCert.getIssuerX500Principal(), span.context());
+                        if (idx < 0) {
+                            final String hostNames = requestedHostNames.stream().collect(Collectors.joining(", "));
+                            final String msg = "could not determine tenant ID from SNI extension";
+                            log.debug("{}: {}", msg, hostNames);
+                            span.log(Map.of("SNI host names", hostNames));
+                            return Future.failedFuture(new ClientErrorException(
```

Comment:
It is true that the RFC allows for multiple host names to be present in the SNI extension. However, in practice, I do not really see this happening because the client libraries simply put the host name being requested into the extension or they provide means to supply a (single) host name to be put into the extension explicitly. I also wonder what the meaning of multiple host names in the extension should be? "I would like to connect to any of these?" In the Hono documentation we will make clear that the adapters consider the first host name found in the extension and will ignore any other elements. That way we have clarity about the semantics and do not need to make multiple tenant look ups based on additional host names provided in the SNI extension. WDYT?

Reply:
It sounds good to include this behavior in the documentation.